### PR TITLE
support proper ValuesSourceConfig resolution of GEO and GEO_SHAPE.

### PR DESCRIPTION
**Pull request by talevy**
This change includes introduction of
Geo and Geo-Shape ValueSourceType and ValueType.

- include missing value resolution
- include ability for aggregations to support either geo_point or geo_shape

### CODE REVIEW

**Review by imotov — COMMENTED** [review]
Left a small suggestion. In general, it makes sense to me, but @polyfractal should probably review it as well since I am not that familiar with some parts of this code.

File: server/src/main/java/org/elasticsearch/index/fielddata/IndexGeoFieldData.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * under the License.
+ */
+
+package org.elasticsearch.index.fielddata;
+
+
+/**
+ * Specialization of {@link IndexFieldData} for geo points.
+ */
+public interface IndexGeoFieldData extends IndexFieldData<AtomicGeoFieldData> {
```
[the remark this answers is not in the document]
Comment (talevy) [reply server/src/main/java/org/elasticsearch/index/fielddata/IndexGeoFieldData.java]: sure! I updated. I'd like to leave the `Geo` in valuesSource until shape field support is introduced

**Review by polyfractal — APPROVED** [review]
LGTM, barring the `missing` parse order issues we were discussing.  Fine with merging now and revisiting after we talk next week, or just waiting.  The rest of the VS stuff is needed anyway and it looks fine to me.

**Comment by talevy** [comment]
bwc test failed because of a specific ML test. will merge and work on updating the geoshape-doc-values branch to latest master

**Pull request merged**